### ISSUE TITLE
Pushing value to standalone `array<non-empty-string, mixed>` variable works, but not when pushing to an object property

### ISSUE THREAD

**Issue by Ocramius** [vimeo/psalm#4842]
Taking following example:

```php
<?php

/** @psalm-var array<non-empty-string, mixed> $a */
$a = [];

$a['packageQuantity'] = 'foo';

echo json_encode($a); // just to suppress unused $a
```

https://psalm.dev/r/2614d62f3a

No errors are raised by psalm (correct).

When we are using an object property, this no longer works:

```php
<?php

class Foo 
{
    /** @psalm-var array<non-empty-string, mixed> */
    public array $bar = [];
}

$foo = new Foo();
$foo->bar['baz'] = 'aaa';
```

```
Psalm output (using commit 2021c81): 

ERROR: PropertyTypeCoercion - 10:1 - $foo->bar expects 'array<non-empty-string, mixed>',  parent type 'array{baz: string(aaa)}<non-empty-string, mixed>' provided
```
https://psalm.dev/r/070bd91263

**Comment by weirdan** [comment]
No longer reproducible: psalm.dev/r/070bd91263
